### PR TITLE
sync with upstream

### LINKED ISSUE #62: Adding Java 7, 12 and 13-ea versions
It would be really handy if we had official Clojure images for those Java versions, mainly for Docker-based CI use. In particular, CircleCI's clojure images can be expanded to have these.

A lot of Clojure projects still aim to support Java versions less than 8, but it becomes increasingly hard to test these. Similarly, being able to test against the latest (and next) stable versions of Java is very useful.

Base images for all of the above are [provided for Docker](https://hub.docker.com/_/openjdk)

### LINKED ISSUE #66: Tools deps out of date
Latest version is `1.10.1.469`

### LINKED ISSUE #67: Provide named Debian variants
Debian Buster has been out since a few months ago and is the current stable version. But Debian Stretch will still be maintained (as `oldstable`) for several months (in theory until 12 months later than Buster official release date).

OpenJDK already provides named Debian variants (e.g., `openjdk:8-stretch`, `openjdk:8-slim-buster[1]`), with `openjdk:8`, `openjdk:8-jre` and `openjdk:8-jdk` defaulting to Debian Stretch as of today (see [2]). But they also warn that "...you'll likely want to specify one of these[3] explicitly to minimize breakage when there are new releases of Debian."

So using `openjdk:8` as the base to to build the Debian-based Clojure images has the risk of suddenly changing Debian distribution when the OpenJDK Docker image maintainers change the Debian release they point to for the `openjdk:8` tag.

I think it would be a good idea to build images for both the Stretch and Buster (and in the general case for `stable` and `oldstable` versions of Debian at that point) and tag them using a similar naming structure than the OpenJDK Docker images (to make it easier to know which image they are based on). That would allow each user to decide whether they want to use one of the other, and migrate from the older one to the when they mean it (as opposed to migrating by simply pulling the same image tag at a later time).

[1] They seem to have change naming patterns for newer Debian releases, as there is no  `stretch` tag having the `slim` part in them, while all `buster` tags have the `slim` part in them.
[2] https://github.com/docker-library/docs/blob/master/openjdk/README.md#shared-tags
[3] They refer to images whose tags may have names like buster or stretch in them. These are the suite code names for releases of Debian and indicate which release the image is based on.

### LINKED ISSUE #68: Clojure v1.10.1 Has Been Released
Can we get an updated image for this?

https://clojure.org/news/2019/06/06/clojure1-10-1

### PR DESCRIPTION
**Pull request by cap10morgan**
OK, here we go (finally).

This sets us up to build on the new upstream openjdk images. The changes didn't end up being too bad once I really dug into it.

Here are the things that this changes and why:

* alpine builds all go away except for JDK 14 EA, but even that might go away when JDK 14 is released as has happened with previous versions. ~~I'm not sure if this means they will be deleted from Docker Hub or what, but we may want to inquire to find out so we can give users a heads up.~~ (UPDATE: They won't be deleted. See @tianon's comments below for details. We should still encourage people to migrate to `slim-buster`). The OpenJDK builds are gone, so I don't think we have any other (good) options here.
* Closes #62 by adding Java 13 and 14 EA builds.
* Closes #67 by adding in `-stretch`, `-slim-buster`, and `-buster` tag suffixes where appropriate.
* OpenJDK 8 and 11 images that don't specify a Debian version will continue to use stretch as they do today to minimize breakage for users.
* OpenJDK 13+ images that don't specify a Debian version will use `slim-buster`. I think this is a good base to default to going forward, and those that want something else can explicitly get it with the new Debian version tag suffixes. Going forward I think the latest slim, stable Debian release makes sense to default to as new openjdk images appear upstream that are based on it.
* The default OpenJDK version if you don't specify one in your Docker tag will now be 11. Going forward I think it makes sense for this to be the latest LTS release. If you want something else, you can explicitly get it with the appropriate tag.
* Closes #66 by providing the latest tools-deps release.
* Updates boot to latest even though we don't have an issue for that yet.
* Closes #68 by pre-bundling Clojure 1.10.1 in the lein images, and implicitly in the tools-deps images since those are tied together.
* Moves the `latest` tag to `openjdk-11-lein-2.9.1-slim-buster`. Going forward I think it makes sense to put this tag on the latest LTS openjdk leiningen image w/ the latest Debian slim stable release. But I could also imagine moving it to `openjdk-13-lein-2.9.1-slim-buster` instead. I guess the question is, should `latest` point at latest LTS Java or latest, latest Java? I don't think it makes sense to point it at an EA Java.
* Simplifies the maintainers list. Let me know if it's now _too_ simple. Just trying to inform users who they should contact for help with an image, but let me know if I got it wrong in any way.

😅 I _think_ that's it?

### CODE REVIEW

**Comment by cap10morgan** [comment]
@tianon If we merge this, will the deleted alpine builds disappear from Docker Hub? If so, on what kind of time frame? Thanks!

**Comment by tianon** [comment]
Removing tags will remove them from the "Supported" section on the Hub readme (and will prevent us from spending cycles rebuilding them on the official build servers), but the tags will still be available to users who want them.  (See https://github.com/docker-library/official-images#library-definition-files for more detail on this.)

**Comment by cap10morgan** [comment]
@tianon Great, thank you!

File: test/docker_clojure/core_test.clj [hunk 1]
```diff
@@ -3,69 +3,111 @@
...
-              :build-tool-version "1.10.1.469"}}
-           (image-variants #{"openjdk:8" "openjdk:11"}
-                           #{"debian" "alpine"}
-                           #{"lein" "boot" "tools-deps"})))))
+    (let [variants (image-variants #{8 11 13 14}
+                                   #{"stretch" "slim-buster" "alpine"}
+                                   {"lein"       "2.9.1"
+                                    "boot"       "2.8.3"
+                                    "tools-deps" "1.10.1.469"})]
+      (is (contains? variants
```
Comment (Quantisan) [inline test/docker_clojure/core_test.clj]: using `(are ...` might read nicer here

**Comment by Quantisan** [comment]
@cap10morgan your decision to pin `latest` to latest LTS Java makes sense to me. Less surprises to users not specifying any tags, as that defaults to `latest`.

**Pull request merged**